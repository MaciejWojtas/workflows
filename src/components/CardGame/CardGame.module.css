.game {
  @apply flex flex-col justify-between flex-1;
}

.game-win {
  @apply absolute;
  top: 15%;
  left: 50%;
  transform: translateX(-50%);
}

.game-counter {
  @apply flex justify-center;
}

.game-cards {
  @apply flex justify-center flex-col flex-1;
}

.game-cards__container {
  @apply flex flex-wrap mx-auto justify-center max-w-sm xl:max-w-full;
}
.game-cards__card {
  @apply mx-2 my-2 xl:my-0 xl:my-0;
  animation: cardThrow 0.3s ease-in-out;
}

@media (min-width: 1280px) {
  .game-cards__card-0 {
    @apply absolute;
    transform: rotate(15deg) translate(-500px, 68px) translateY(-50%);
  }
  .game-cards__card-1 {
    z-index: 1;
    @apply md:absolute;
    transform: var(--rotate-1, rotate(7.5deg)) var(--translate-1, translate(-250px, 16px))
      translateY(-50%);
  }
  .game-cards__card-2 {
    z-index: 2;
    @apply md:absolute;
    transform: var(--rotate-2, rotate(0deg)) var(--translate-2, translate(0)) translateY(-50%);
  }
  .game-cards__card-3 {
    z-index: 3;
    @apply md:absolute;
    transform: var(--rotate-3, rotate(-7.5deg)) var(--translate-3, translate(250px, 16px))
      translateY(-50%);
  }
  .game-cards__card-4 {
    z-index: 4;
    @apply md:absolute;
    transform: var(--rotate-4, rotate(-15deg)) var(--translate-4, translate(500px, 68px))
      translateY(-50%);
  }
}

.game-deal {
  @apply min-w-full flex flex-col justify-end justify-center  h-60 xl:h-auto xl:mb-12;
}
.game-lose {
  @apply mb-16 text-white text-2xl xl:text-4xl max-w-lg mx-auto;
}

.game-deal__button {
  @apply uppercase mx-auto;
}

.game-reset-btn {
  @apply mx-auto;
}

.game-reset {
  @apply min-w-full flex justify-center xl:justify-end;
}

.game-reset__button {
  @apply xl:mr-0;
}

.body-inner {
  @apply text-center;
}

@keyframes cardThrow {
  0% {
    --translate-1: translate(-500px, 68px);
    --translate-2: translate(-500px, 68px);
    --translate-3: translate(-500px, 68px);
    --translate-4: translate(-500px, 68px);
    --rotate-1: rotate(15deg);
    --rotate-2: rotate(15deg);
    --rotate-3: rotate(15deg);
    --rotate-4: rotate(15deg);
  }

  20% {
    --translate-1: translate(-375px, 34px);
    --translate-2: translate(-375px, 34px);
    --translate-3: translate(-375px, 34px);
    --translate-4: translate(-375px, 34px);
    --rotate-1: rotate(10.5deg);
    --rotate-2: rotate(10.5deg);
    --rotate-3: rotate(10.5deg);
    --rotate-4: rotate(10.5deg);
  }

  40% {
    --translate-2: translate(-135px, 8px);
    --translate-3: translate(-135px, 8px);
    --translate-4: translate(-135px, 8px);
    --rotate-2: rotate(5deg);
    --rotate-3: rotate(5deg);
    --rotate-4: rotate(5deg);
  }

  60% {
    --translate-3: translate(135px, 8px);
    --translate-4: translate(135px, 8px);
    --rotate-3: rotate(-5deg);
    --rotate-4: rotate(-5deg);
  }

  80% {
    --translate-4: translate(375px, 34px);
    --rotate-4: rotate(-10.5deg);
  }

  100% {
  }
}
