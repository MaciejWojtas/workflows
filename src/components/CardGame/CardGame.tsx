import React, { FunctionComponent } from 'react';
import Confetti from 'react-confetti';
import { ReactComponent as Winner } from '../../assets/img/winner.svg';
import useCardDeck from '../../hooks/useCardDeck';
import Card from './Card/Card';
import Counter from './Counter/Counter';
import styles from './CardGame.module.css';

const CardGame: FunctionComponent = () => {
  const { deal, makeDeal, reset, cardsLeft, acesLeft } = useCardDeck();

  const win = cardsLeft === 0 && acesLeft === 0;
  const loss = cardsLeft !== 0 && acesLeft === 0;
  const hasEndedGame = loss || win;

  return (
    <div className={styles.game}>
      {win ? <Confetti gravity={0.5} numberOfPieces={100} recycle={false} /> : null}
      {win ? (
        <div className={styles['game-win']}>
          <Winner />
        </div>
      ) : null}
      <div className={styles['game-counter']}>
        <Counter cardsLeft={cardsLeft} acesLeft={acesLeft} />
      </div>
      <div className={styles['game-cards']}>
        <div className={styles['game-cards__container']}>
          {deal.map((card, index) => (
            <Card
              className={`${styles['game-cards__card']} ${
                deal.length === 5 ? styles[`game-cards__card-${index}`] : ''
              }`}
              key={`${card.rank} ${card.suit}`}
              card={card}
            />
          ))}
        </div>
      </div>

      <div className={`${styles['game-deal']}`}>
        {loss ? (
          <span className={`${styles['game-lose']} font-secondary`}>
            You lose.
            <br /> Better luck next time!
          </span>
        ) : null}
        {hasEndedGame ? (
          <button
            className={`${styles['game-reset-btn']} btn-secondary`}
            type="button"
            onClick={reset}
          >
            Play Again
          </button>
        ) : (
          <button
            className={`${styles['game-deal__button']} btn-primary`}
            type="button"
            onClick={makeDeal}
          >
            Deal
          </button>
        )}
      </div>
      {hasEndedGame ? null : (
        <div className={styles['game-reset']}>
          <button
            className={`${styles['game-reset-btn']} ${styles['game-reset__button']} btn-secondary`}
            type="button"
            onClick={reset}
          >
            Reset
          </button>
        </div>
      )}
    </div>
  );
};

export default CardGame;
