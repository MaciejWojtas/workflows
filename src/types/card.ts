export const ranks = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'] as const;
export const suits = ['C', 'D', 'H', 'S'] as const;

export type CardRank = typeof ranks[number];
export type CardSuit = typeof suits[number];

export type Card = {
  rank: CardRank;
  suit: CardSuit;
};
