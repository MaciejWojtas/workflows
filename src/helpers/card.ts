import { ReactComponent as Clover } from '../assets/img/cards/Clover.svg';
import { ReactComponent as Diamond } from '../assets/img/cards/Diamond.svg';
import { ReactComponent as Heart } from '../assets/img/cards/Heart.svg';
import { ReactComponent as Spade } from '../assets/img/cards/Spade.svg';
import { Card } from '../types/card';

export const isAce = (card: Card): boolean => card.rank === 'A';

const suitImages = {
  C: Clover,
  D: Diamond,
  H: Heart,
  S: Spade,
};

const redSuits = ['H', 'D'];

export const getSuitImage = (card: Card) => suitImages[card.suit];
export const isRedCard = (card: Card) => redSuits.includes(card.suit);
