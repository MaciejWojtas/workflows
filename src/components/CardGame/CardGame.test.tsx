import React from 'react';
import renderer from 'react-test-renderer';
import { cloverCard, spadeCard, diamondCard, heartCard, card } from '../../tests/mocks/cards';
import { Card } from '../../types/card';
import CardGame from './CardGame';

let mockDeal: Card[] = [];
let mockCardsLeft = 52;
let mockAcesLeft = 4;

jest.mock('../../hooks/useCardDeck', () => ({
  __esModule: true,
  default: () => ({
    deal: mockDeal,
    cardsLeft: mockCardsLeft,
    acesLeft: mockAcesLeft,
    makeDeal: jest.fn(),
    reset: jest.fn(),
  }),
}));

it('renders correctly ', () => {
  const tree = renderer.create(<CardGame />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders correctly after deal of 5', () => {
  mockDeal = [cloverCard, spadeCard, diamondCard, heartCard, card];
  mockCardsLeft = 47;
  mockAcesLeft = 3;
  const tree = renderer.create(<CardGame />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders correctly when lose', () => {
  mockDeal = [cloverCard, spadeCard, diamondCard, heartCard, card];
  mockCardsLeft = 47;
  mockAcesLeft = 0;
  const tree = renderer.create(<CardGame />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders correctly after last deal of 2 and win', () => {
  mockDeal = [cloverCard, card];
  mockCardsLeft = 0;
  mockAcesLeft = 0;
  const tree = renderer.create(<CardGame />).toJSON();

  expect(tree).toMatchSnapshot();
});
