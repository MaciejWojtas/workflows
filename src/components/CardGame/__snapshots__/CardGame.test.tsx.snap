// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly  1`] = `
<div
  className="game"
>
  <div
    className="game-counter"
  >
    <div
      className="counter"
    >
      <span
        className="counter__number font-secondary"
      >
        52
      </span>
      <span
        className="counter__text font-secondary"
      >
        Cards Left
      </span>
      <div
        className="counter-aces font-secondary"
      >
        <span
          className="counter-aces__ace"
        >
          A
        </span>
        <span
          className="counter-aces__ace"
        >
          A
        </span>
        <span
          className="counter-aces__ace"
        >
          A
        </span>
        <span
          className="counter-aces__ace"
        >
          A
        </span>
      </div>
    </div>
  </div>
  <div
    className="game-cards"
  >
    <div
      className="game-cards__container"
    />
  </div>
  <div
    className="game-deal"
  >
    <button
      className="game-deal__button btn-primary"
      onClick={[MockFunction]}
      type="button"
    >
      Deal
    </button>
  </div>
  <div
    className="game-reset"
  >
    <button
      className="game-reset-btn game-reset__button btn-secondary"
      onClick={[MockFunction]}
      type="button"
    >
      Reset
    </button>
  </div>
</div>
`;

exports[`renders correctly after deal of 5 1`] = `
<div
  className="game"
>
  <div
    className="game-counter"
  >
    <div
      className="counter"
    >
      <span
        className="counter__number font-secondary"
      >
        47
      </span>
      <span
        className="counter__text font-secondary"
      >
        Cards Left
      </span>
      <div
        className="counter-aces font-secondary"
      >
        <span
          className="counter-aces__ace"
        >
          A
        </span>
        <span
          className="counter-aces__ace"
        >
          A
        </span>
        <span
          className="counter-aces__ace"
        >
          A
        </span>
      </div>
    </div>
  </div>
  <div
    className="game-cards"
  >
    <div
      className="game-cards__container"
    >
      <div
        className="card game-cards__card game-cards__card-0"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank "
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Clover.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Clover.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-1"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank "
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Spade.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Spade.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-2"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank card-desc__rank--red"
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Diamond.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Diamond.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-3"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank card-desc__rank--red"
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Heart.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Heart.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-4"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank card-desc__rank--red"
          >
            10
          </span>
          <svg
            className="card-desc__img"
          >
            Diamond.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Diamond.svg
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="game-deal"
  >
    <button
      className="game-deal__button btn-primary"
      onClick={[MockFunction]}
      type="button"
    >
      Deal
    </button>
  </div>
  <div
    className="game-reset"
  >
    <button
      className="game-reset-btn game-reset__button btn-secondary"
      onClick={[MockFunction]}
      type="button"
    >
      Reset
    </button>
  </div>
</div>
`;

exports[`renders correctly after last deal of 2 and win 1`] = `
<div
  className="game"
>
  <canvas
    height={768}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 2,
      }
    }
    width={1024}
  />
  <div
    className="game-win"
  >
    <svg>
      winner.svg
    </svg>
  </div>
  <div
    className="game-counter"
  >
    <div
      className="counter"
    >
      <span
        className="counter__number font-secondary"
      >
        0
      </span>
      <span
        className="counter__text font-secondary"
      >
        Cards Left
      </span>
      <div
        className="counter-aces font-secondary"
      />
    </div>
  </div>
  <div
    className="game-cards"
  >
    <div
      className="game-cards__container"
    >
      <div
        className="card game-cards__card "
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank "
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Clover.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Clover.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card "
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank card-desc__rank--red"
          >
            10
          </span>
          <svg
            className="card-desc__img"
          >
            Diamond.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Diamond.svg
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="game-deal"
  >
    <button
      className="game-reset-btn btn-secondary"
      onClick={[MockFunction]}
      type="button"
    >
      Play Again
    </button>
  </div>
</div>
`;

exports[`renders correctly when lose 1`] = `
<div
  className="game"
>
  <div
    className="game-counter"
  >
    <div
      className="counter"
    >
      <span
        className="counter__number font-secondary"
      >
        47
      </span>
      <span
        className="counter__text font-secondary"
      >
        Cards Left
      </span>
      <div
        className="counter-aces font-secondary"
      />
    </div>
  </div>
  <div
    className="game-cards"
  >
    <div
      className="game-cards__container"
    >
      <div
        className="card game-cards__card game-cards__card-0"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank "
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Clover.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Clover.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-1"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank "
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Spade.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Spade.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-2"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank card-desc__rank--red"
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Diamond.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Diamond.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-3"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank card-desc__rank--red"
          >
            A
          </span>
          <svg
            className="card-desc__img"
          >
            Heart.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Heart.svg
          </svg>
        </div>
      </div>
      <div
        className="card game-cards__card game-cards__card-4"
      >
        <div
          className="card-desc"
        >
          <span
            className="card-desc__rank card-desc__rank--red"
          >
            10
          </span>
          <svg
            className="card-desc__img"
          >
            Diamond.svg
          </svg>
        </div>
        <div
          className="card-suit"
        >
          <svg
            className="card-suit__img"
          >
            Diamond.svg
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="game-deal"
  >
    <span
      className="game-lose font-secondary"
    >
      You lose.
      <br />
       Better luck next time!
    </span>
    <button
      className="game-reset-btn btn-secondary"
      onClick={[MockFunction]}
      type="button"
    >
      Play Again
    </button>
  </div>
</div>
`;
